### ISSUE TITLE
case-insensitive regex bug with UTF8-flagged strings

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12429]
Migrated from [rt.perl.org#114982](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=114982) (status was 'resolved')

Searchable as RT114982$

**Comment by p5pRT** [comment]
### From peter@peknet.com

#### Created by karman@cpan.org



Failing test below&#8203;:

\#\!/usr/bin/env perl
use strict;
use warnings;
use Test&#8203;::More tests => 2;

my $str     = ' Foo ';    \# uppercase breaks 2nd regex
my $pattern = 'foo';
utf8&#8203;::upgrade\($str\);
utf8&#8203;::upgrade\($pattern\);

my $re          = qr/$pattern/i;
my $re\_optional = qr/\(?i&#8203;:\.?\)$pattern/i;

like\( $str\, $re\,          "re" \);
like\( $str\, $re\_optional\, "re\_optional" \);

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=high

Site configuration information for perl 5.17.4:

Configured by pkarman at Thu Sep 20 12:22:16 CDT 2012.

Summary of my perl5 (revision 5 version 17 subversion 4) configuration:

   Platform:
     osname=linux, osvers=2.6.18-308.1.1.el5, archname=x86_64-linux
     uname='linux pkarmanpc.mpr.org 2.6.18-308.1.1.el5 #1 smp wed mar 7 
04:16:51 est 2012 x86_64 x86_64 x86_64 gnulinux '
     config_args='-des -Dprefix=/usr/local/perl/5.17.4 -Dusedevel'
     hint=recommended, useposix=true, d_sigaction=define
     useithreads=undef, usemultiplicity=undef
     useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
     use64bitint=define, use64bitall=define, uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector 
-I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
     optimize='-O2',
     cppflags='-fno-strict-aliasing -pipe -fstack-protector 
-I/usr/local/include'
     ccversion='', gccversion='4.1.2 20080704 (Red Hat 4.1.2-52)', 
gccosandvers=''
     intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
     ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', 
lseeksize=8
     alignbytes=8, prototype=define
   Linker and Libraries:
     ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
     libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64 /usr/local/lib64
     libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lc
     perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
     libc=/lib/libc-2.5.so, so=so, useshrplib=false, libperl=libperl.a
     gnulibc_version='2.5'
   Dynamic Linking:
     dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
     cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib 
-fstack-protector'

Locally applied patches:



@INC for perl 5.17.4:
     /usr/local/perl/5.17.4/lib/site_perl/5.17.4/x86_64-linux
     /usr/local/perl/5.17.4/lib/site_perl/5.17.4
     /usr/local/perl/5.17.4/lib/5.17.4/x86_64-linux
     /usr/local/perl/5.17.4/lib/5.17.4
     .


Environment for perl 5.17.4:
     HOME=/home/pkarman
     LANG=en_US.UTF-8
     LANGUAGE (unset)
     LD_LIBRARY_PATH=/home/pkarman/sw/lib:
     LOGDIR (unset)
 
PATH=/home/pkarman/bin:/opt/perl/bin:/opt/pij/bin:/opt/subversion/bin:/opt/drupal/bin:/usr/kerberos/bin:/usr/local/bin:/bin:/usr/bin
     PERL_BADLANG (unset)
     SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Sep 20 10&#8203;:46&#8203;:02 2012\, peter@&#8203;peknet\.com wrote&#8203;:
> This is a bug report for perl from karman@&#8203;cpan\.org\,
> generated with the help of perlbug 1\.39 running under perl 5\.17\.4\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> Failing test below&#8203;:
> 
> \#\!/usr/bin/env perl
> use strict;
> use warnings;
> use Test&#8203;::More tests => 2;
> 
> my $str     = ' Foo ';    \# uppercase breaks 2nd regex
> my $pattern = 'foo';
> utf8&#8203;::upgrade\($str\);
> utf8&#8203;::upgrade\($pattern\);
> 
> my $re          = qr/$pattern/i;
> my $re\_optional = qr/\(?i&#8203;:\.?\)$pattern/i;
> 
> like\( $str\, $re\,          "re" \);
> like\( $str\, $re\_optional\, "re\_optional" \);

b0573d8bdb309bce282466c30d1092bc2d421ed7 is the first bad commit
commit b0573d8bdb309bce282466c30d1092bc2d421ed7
Author&#8203;: Karl Williamson \<public@&#8203;khwilliamson\.com>
Date&#8203;:   Sun Apr 29 09&#8203;:33&#8203;:19 2012 \-0600

&nbsp;    regexec\.c&#8203;: Use foldcase instead of lower/upper
&nbsp;    
&nbsp;    The way Perl deals with folds of ASCII\-range characters\, is to have an
&nbsp;    array that maps to the opposite case character\.  For example\, fold\[a\] is
&nbsp;    A\, and fold\[A\] is a\.  To see if a character matches a pattern character
&nbsp;    case\-insensitively\, you compare it against the pattern character itself\,
&nbsp;    and if that fails\, against the fold\[pattern character\]\.
&nbsp;    
&nbsp;    Unfortunately that breaks down outside of ASCII\, where the folds are not
&nbsp;    necessarily one\-to\-one\.  This code is a vestige of that original
&nbsp;    paradim\, and I'm converting it to use the more modern\.  The apparent
&nbsp;    reason it doesn't fail in fold\_grind\.t is that it is for CURLY nodes\,
&nbsp;    and CURLY is not currently set on EXACTish nodes unless the node
&nbsp;    contains a single UTF\-8 invariant character\.   Effectively\, that means
&nbsp;    a character in the ASCII range\.
&nbsp;    
&nbsp;    I think parts of this could be simplified further\, but I don't
&nbsp;    understand the whole context of this code to undertake that\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @khwilliamson

On 09/20/2012 01&#8203;:53 PM\, Father Chrysostomos via RT wrote&#8203;:
> On Thu Sep 20 10&#8203;:46&#8203;:02 2012\, peter@&#8203;peknet\.com wrote&#8203;:
>> This is a bug report for perl from karman@&#8203;cpan\.org\,
>> generated with the help of perlbug 1\.39 running under perl 5\.17\.4\.
>>
>>
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> \[Please describe your issue here\]
>>
>> Failing test below&#8203;:
>>
>> \#\!/usr/bin/env perl
>> use strict;
>> use warnings;
>> use Test&#8203;::More tests => 2;
>>
>> my $str     = ' Foo ';    \# uppercase breaks 2nd regex
>> my $pattern = 'foo';
>> utf8&#8203;::upgrade\($str\);
>> utf8&#8203;::upgrade\($pattern\);
>>
>> my $re          = qr/$pattern/i;
>> my $re\_optional = qr/\(?i&#8203;:\.?\)$pattern/i;
>>
>> like\( $str\, $re\,          "re" \);
>> like\( $str\, $re\_optional\, "re\_optional" \);
>
> b0573d8bdb309bce282466c30d1092bc2d421ed7 is the first bad commit
> commit b0573d8bdb309bce282466c30d1092bc2d421ed7
> Author&#8203;: Karl Williamson \<public@&#8203;khwilliamson\.com>
> Date&#8203;:   Sun Apr 29 09&#8203;:33&#8203;:19 2012 \-0600
>
>      regexec\.c&#8203;: Use foldcase instead of lower/upper
>
>      The way Perl deals with folds of ASCII\-range characters\, is to have an
>      array that maps to the opposite case character\.  For example\, fold\[a\] is
>      A\, and fold\[A\] is a\.  To see if a character matches a pattern character
>      case\-insensitively\, you compare it against the pattern character itself\,
>      and if that fails\, against the fold\[pattern character\]\.
>
>      Unfortunately that breaks down outside of ASCII\, where the folds are not
>      necessarily one\-to\-one\.  This code is a vestige of that original
>      paradim\, and I'm converting it to use the more modern\.  The apparent
>      reason it doesn't fail in fold\_grind\.t is that it is for CURLY nodes\,
>      and CURLY is not currently set on EXACTish nodes unless the node
>      contains a single UTF\-8 invariant character\.   Effectively\, that means
>      a character in the ASCII range\.
>
>      I think parts of this could be simplified further\, but I don't
>      understand the whole context of this code to undertake that\.
>

If we want this regression removed for 5\.16\.2\, I propose simply 
reverting this patch\.

However\, it turns out that this is showing a whole new class of folding 
bugs that I overlooked when trying to fix things earlier\.  The fixes for 
these are too complicated for a dot release\.  Reverting the patch just 
gets us back to bug parity with earlier releases\, and a more complete 
fix can come in 5\.18

**Comment by p5pRT** [comment]
### From peter@peknet.com

karl williamson via RT wrote on 9/21/12 12&#8203;:26 PM&#8203;:

> 
> If we want this regression removed for 5\.16\.2\, I propose simply 
> reverting this patch\.

Not sure if I am following the release management logic here\, but just wanted to
note that the bug does not appear in 5\.16\.1 \(i\.e\. the tests pass there\)\.


\-\- 
Peter Karman  \.  http&#8203;://peknet\.com/  \.  peter@&#8203;peknet\.com

**Comment by p5pRT** [comment]
### From @khwilliamson

On 09/21/2012 12&#8203;:07 PM\, Peter Karman wrote&#8203;:
> karl williamson via RT wrote on 9/21/12 12&#8203;:26 PM&#8203;:
>
>>
>> If we want this regression removed for 5\.16\.2\, I propose simply
>> reverting this patch\.
>
> Not sure if I am following the release management logic here\, but just wanted to
> note that the bug does not appear in 5\.16\.1 \(i\.e\. the tests pass there\)\.
>
>

Oh\.  I just looked at the date of the commit\, and assumed it made it 
into 5\.16\.  But it didn't\, so I will fix this in some 5\.17 load\.  Thanks 
for the report\.

**Comment by p5pRT** [comment]
### From @khwilliamson

This is fixed by this&#8203;:

commit c74f6de970ef0f0eb8ba43b1840fde0cf5a45497
 Author&#8203;: Karl Williamson \<public@&#8203;khwilliamson\.com>
 Date&#8203;:   Wed Oct 3 22&#8203;:17&#8203;:19 2012 \-0600
 
\*    PATCH&#8203;: \[perl \#114982\]&#8203;: case\-insensitive regex bug with UTF8\-flagged
strings
&nbsp;     
&nbsp;     It turns out that this whole area in regexec\.c has never worked
&nbsp;     properly\, we just didn't have test coverage for it\.  Indeed\, the
portion
&nbsp;     that deals with CURLYM had never been updated to include UTF\-8\!
&nbsp;     The failure rate for the new tests added by this commit on the blead
&nbsp;     that existed just prior to this commit was 96%\.  \(4% pass\)
&nbsp;     
&nbsp;     This commit refactors the part of regexec\.c that deals with quantifiers
&nbsp;     and case insensitive matching\, and adds comprehensive tests\.
&nbsp;     
&nbsp;     Consider a regex of the form A\*B\.  How do we know how many of A\* to
&nbsp;     match?  The answer is we don't\, until we try B\.  But if B begins with
&nbsp;     text\, we can easily rule out places where the next thing isn't a B\.
&nbsp;     Only if the next character \(following where we are\) is the first
&nbsp;     character of B\, is this a potential match; otherwise it isn't\, so there
&nbsp;     is no need to even try B\.
&nbsp;     
&nbsp;     Code existed to do this\, but it was wrong in many ways\.  If B can match
&nbsp;     case\-insensitively under /i\, then there may be two or more possible
&nbsp;     characters that could begin B\.  The previous code assumed there was a
&nbsp;     max of two\.  It didn't adequately account for the differences between
&nbsp;     /a\, /d\, /l\, and /u; and as I mentioned above\, in one place it didn't
&nbsp;     know that there was such a thing as UTF\-8\.  Also\, it assumed that all
&nbsp;     code points can fit into an I32\.
&nbsp;     
&nbsp;     To do this right takes quite a bit more intelligence then was there;
&nbsp;     and it needs to be done in two different places\.  So\, this commit
&nbsp;     extracts out the code into a single subroutine\, heavily modified to
&nbsp;     account for the various oversights that were there previously\.
&nbsp;     
&nbsp;     The pre\-existing infrastructure of fold\_grind\.t allowed the tests to be
&nbsp;     added easily\.  I did not add tests for the above\-I32 problems\, as it
&nbsp;     turns out that there are other bugs on these\, \[perl \#115166\]

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'resolved'
